### PR TITLE
Implement fail with scalars

### LINKED ISSUE #203: Support scalars in fail instruction
`fail` instruction supports only literals and `%last_error%`, but in certain situations it's neccessary to "throw" a scalar, f.e.

```wat
(xor
  (call "" ("" "") []) ;; will thrown an error
  (seq
    (ap %last_error% some_scalar) ;; save this scalar to handle it later
    ... ;; try to handle it
    (fail some_scalar) ;; throw it again
  )
)
```

To make error-flow unifirm, scalars in `fail` must be object type and have at least two fields `message` and `error_code`.

### PR DESCRIPTION
**Pull request by mikevoronov**
Resolves #203.

### CODE REVIEW

File: air/src/execution_step/air/fail.rs [hunk 1]
```diff
@@ -45,25 +49,32 @@ impl<'i> super::ExecutableInstruction<'i> for Fail<'i> {
     }
 }
 
+fn fail_with_scalar<'i>(scalar: &ast::ScalarWithLambda<'i>, exec_ctx: &mut ExecutionCtx<'i>) -> ExecutionResult<()> {
+    let (value, mut tetraplet) = resolve_ast_scalar_wl(scalar, exec_ctx)?;
+    // tetraplets always have one elements here and it'll be refactored after boxed value
```
Comment (ValeryAntopol) [inline air/src/execution_step/air/fail.rs:54]:
```suggestion
    // tetraplets always have one element here and it'll be refactored after boxed value
```

File: air/src/execution_step/air/fail.rs [hunk 2]
```diff
@@ -45,25 +49,32 @@ impl<'i> super::ExecutableInstruction<'i> for Fail<'i> {
...
+    let rtetraplet = tetraplet.remove(0);
```
Comment (ValeryAntopol) [inline air/src/execution_step/air/fail.rs:55]: Maybe just shadow old `tetraplet`? Or make a meaningful name, if shadowing is not possible.

File: air/src/execution_step/air/par.rs [hunk 3]
```diff
@@ -38,10 +38,10 @@ impl<'i> ExecutableInstruction<'i> for Par<'i> {
         trace_to_exec_err!(trace_ctx.meet_par_start(), self)?;
 
         // execute a left subtree of par
-        let left_result = execute_subtree(&self, exec_ctx, trace_ctx, &mut completeness_updater, SubtreeType::Left)?;
+        let left_result = execute_subtree(self, exec_ctx, trace_ctx, &mut completeness_updater, SubtreeType::Left)?;
 
         // execute a right subtree of par
-        let right_result = execute_subtree(&self, exec_ctx, trace_ctx, &mut completeness_updater, SubtreeType::Right)?;
+        let right_result = execute_subtree(self, exec_ctx, trace_ctx, &mut completeness_updater, SubtreeType::Right)?;
```
Comment (ValeryAntopol) [inline air/src/execution_step/air/par.rs:44]: How does it work? We move `self` twice here

File: air/src/execution_step/errors/catchable_errors.rs [hunk 4]
```diff
@@ -77,6 +78,10 @@ pub enum CatchableError {
     /// Errors occurred while insertion of a value inside stream that doesn't have corresponding generation.
     #[error("stream {0:?} doesn't have generation with number {1}, probably a supplied to the interpreter data is corrupted")]
     StreamDontHaveSuchGeneration(Stream, usize),
+
+    /// This error type is produced by a fail instruction that tries to throw a scalar that have inappropriate type.
+    #[error(transparent)]
+    LastErrorObjectError(#[from] LastErrorObjectError),
```
Comment (ValeryAntopol) [inline air/src/execution_step/errors/catchable_errors.rs:84]: maybe `InvalidLastErrorObjectError`?

File: air/tests/test_module/instructions/fail.rs [hunk 5]
```diff
@@ -58,13 +58,15 @@ fn fail_with_literals() {
...
+    let init_peer_id = "init_peer_id";
+    let result = call_vm!(vm, init_peer_id, script, "", "");
 
     let expected_error = CatchableError::UserError {
         error: rc!(json!( {
         "error_code": 1337i64,
         "instruction": "fail 1337 error message",
         "message": "error message",
+            "peer_id": init_peer_id,
         })),
```
Comment (ValeryAntopol) [inline air/tests/test_module/instructions/fail.rs:70]:
```suggestion
        "peer_id": init_peer_id,
        })),
```

File: air/src/execution_step/execution_context/last_error/last_error_definition.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        fields
+            .get(field_name)
+            .ok_or_else(|| LastErrorObjectError::ScalarMustContainField {
+                scalar: scalar.clone(),
+                field_name,
+            })
+    };
+
+    check_field(ERROR_CODE_FIELD_NAME)?;
+    check_field(MESSAGE_FIELD_NAME)?;
```
Comment (ValeryAntopol) [inline air/src/execution_step/execution_context/last_error/last_error_definition.rs:74]: Lets also check types of these fields

File: air/src/execution_step/air/par.rs (hunk 3 above)
Comment (mikevoronov) [reply air/src/execution_step/air/par.rs:44]: It's ok since it's a reference.

**Review by ValeryAntopol — APPROVED** [review]

**Pull request merged**